db-engine/src/main/java/com/splicemachine/db/iapi/sql/dictionary/fk/Graph.java

```diff
@@ -0,0 +1,181 @@
+                    return;
+                }
+            }
+        }
+        // cycle is unbreakable, bail out
+        StringBuilder sb = new StringBuilder();
+        sb.append("adding the constraint between ").append(
+                getName(p.get(0))).append(" and ").append(getName(p.get(p.size()-1))).append(" would cause the following illegal delete action cascade cycle");
+        for(int v : p) {
+            sb.append(" ").append(getName(v));
+        }
+        throw StandardException.newException(SQLState.LANG_DELETE_RULE_VIOLATION,
+                                             newConstraintName,
+                                             sb.toString());
+    }
+
+    public String getName(int index) {
+        for (Map.Entry<String, Integer> entry : vertexIndex.entrySet()) {
+            if (entry.getValue() == index) {
+                return entry.getKey();
+            }
+        }
+        throw new IllegalArgumentException("could not find name for item at the index: " + index);
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("digraph {").append("\n");
+        int i = 0;
+        for (EdgeNode edge : edgeNodes) {
+            EdgeNode next = edge;
+            while (next != null) {
+                sb.append("\t").append(getName(i)).append(" -> ").append(getName(next.y)).append("[label=\"").append(next.type.toString()).append("\"];").append("\n");
+                next = next.next;
+            }
+            i++;
+        }
+        sb.append("}").append("\n");
+        return sb.toString();
+    }
+
+    public EdgeNode getEdge(int v) {
+        return edgeNodes.get(v);
+    }
+
+    public int getVertexCount() {
+        return vertexIndex.size();
+
+    }
+
+    public EdgeNode.Type getEdgeType(int from, int to) {
+        EdgeNode edgeNode = getEdge(from);
+        while(true) {
+            if(edgeNode == null) {
+                throw new IllegalArgumentException("no edge between " + getName(from) + " and " + getName(to));
+            }
+            if(edgeNode.y == to) {
+                return edgeNode.type;
+            }
+            edgeNode = edgeNode.next;
+        }
+    }
+
+    public void removeEdge(int from, int to) {
+        EdgeNode edgeNode = getEdge(from);
+        if(edgeNode == null) {
+            throw new IllegalArgumentException("no edge between " + getName(from) + " and " + getName(to));
+        }
+        if(edgeNode.y == to) {
+            edgeNodes.set(from, null);
+        } else {
+            EdgeNode previous = edgeNode;
+            edgeNode = edgeNode.next;
+            while(true) {
+                if(edgeNode == null) {
+                    throw new IllegalArgumentException("no edge between " + getName(from) + " and " + getName(to));
+                }
+                if(edgeNode.y == to) {
+                    previous.next = edgeNode.next;
+                    edgeNode.next = null; // bye, GC.
```

Comment:
`edgeNode.next = null; // bye, GC.`
this line is superflous. the thing you want to clean up is edgeNode, not edgeNode.next. edgeNode.next will survive in previous.next. at this point, no one should be pointing to edgeNode anymore, so no need for this line.

Reply:
Good point, but it seems it could help generational GC as explained [here](https://stackoverflow.com/a/6936214/337194) and [here](https://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/util/LinkedList.java#l450).
What do you think?

Reply:
woah hm yeah i didn't think about that! a GC hack, interesting. 👍